### PR TITLE
Cfg error code tests

### PR DESCRIPTION
**Pull request by kevinhartman**
Changes include:
- Cfg error code tests for invalid size and unknown block ID
- result.h and bitfield.h along with some of their dependencies

<b>Note:</b> I removed ResultVal from result.h as it doesn't really make sense client side, and iirc requires more dependencies. I also removed some includes from common.h to avoid bringing in unneeded dependencies.

### CODE REVIEW

**Comment by kevinhartman** [comment]
Actually I think this warrants a discussion. Theoretically, instead of moving utility files like this into the test library and chopping them up, we could register Citra as a git submodule. We'd really need a more robust build system for hwtests though. Reply here or highlight me (peachum) on IRC.

File: source/main.cpp [hunk 1]
```diff
@@ -4,12 +4,13 @@
...
 #include "tests/fs/fs.h"
 #include "tests/cpu/cputests.h"
+#include "tests/cfg/cfg.h"
 #include "tests/kernel/kernel.h"
 
 static unsigned int test_counter = 0;
 static TestCaller tests[] = {
     FS::TestAll,
-    CPU::Integer::TestAll,
+    Cfg::TestAll,
```
Comment (Normmatt) [inline source/main.cpp]: You removed the CPU:Integer tests.

Comment (kevinhartman) [inline source/main.cpp]: Thanks, fixed.

**Comment by Subv** [comment]
I think adding Citra as a submodule is overkill. If anything, we could simply move Common to its own repo

File: source/tests/cfg/cfg.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+    static bool TestCfgBlockNotFoundRC() {
+        const u32 blkID = INVALID_ID.id;
+        const u32 size = INVALID_ID.size;
+        std::array<u8, size> outData;
+
+        Result result = CFGU_GetConfigInfoBlk2(size, blkID, &outData[0]);
+
+        if ((u32)result != RC_BAD_ID.raw)
```
Comment (Subv) [inline source/tests/cfg/cfg.cpp]: Shouldn't you use SoftAssert for this?

**Comment by neobrain** [comment]
I'm seconding including citra as a submodule. The GPU tests will need the `pica.h` and (maybe) `gpu.h` headers anyway, and I think it would be annoying to manually keep these in sync. If anyone is concerned about wasted disk space, they can simply not checkout the submodule and instead point CMake to a separate checkout of the citra repository.

**Comment by archshift** [comment]
Why are you including all of common, common_funcs, etc anyway...?

**Comment by archshift** [comment]
Can you get rid of both the `common.h` and `common_funcs.h` you included? Basically no parts of those files apply to hwtests.

**Comment by archshift** [comment]
As for submoduling Citra, I think that's definitely not a great idea until Citra has been made more modular (across repositories).

**Comment by kevinhartman** [comment]
I don't think we'd ever split Citra up into multiple projects-- submodules are pretty annoying, and it's nice to avoid them. I think the best thing to do would be to include Citra as a submodule and refactor/split the `common` files (in Citra) so that they have more specific utility. We probably *could* add Citra as a submodule now, but I don't have time to set up a new build system. For now, I trimmed out everything besides result.h and bitfield.h.

File: source/tests/cfg/cfg.cpp (hunk 2 above)
Comment (kevinhartman) [inline source/tests/cfg/cfg.cpp]: Yeah I guess. Didn't realize that existed, though I can't say I'm too crazy about this being a macro that returns false. Imagine using that inside a lambda or something. Why didn't the calling test fail? Oh this isn't an exception at all.

**Comment by neobrain** [comment]
As I mentioned before on IRC, the options that we have are basically

1) include citra as a submodule and call it a day
2) work towards modularizing citra and make the core headers a submodule
3) copy over the relevant files
4) redefine the relevant structures in new headers

(4) is something that I hope we can agree on is just not worth the additional effort.
(3) is annoying because it imposes a maintainence burden: Whenever a Citra header changes, we should keep the hwtests copy in sync. If we don't do so, we sooner or later run into the risk of encountering stupid bugs which are fixed in Citra but not in hwtests. It's particular annoying for me since bitfield.h is already spread over enough projects already.
(2) sounds like it would result either in a complete separation into src/ and include/ folders or in a quite artificial "core headers" module. I find both of these options annoying to deal with outside of an IDE.
(1) might be considered including 98% of unused stuff, but what does it really matter? It gets us what we want and disk space is cheap. If you disagree with the latter point, then you'll eventually be able to point CMake to a local Citra checkout and omit the submodule.

In any case, we should agree on a solution ASAP because this problem will become worse in the future and cleaning up the mess that hwtests might have become by then will not be fun.

**Comment by kevinhartman** [comment]
@neobrain I agree with your conclusion. And yeah, I guess things wouldn't start to get messy unless we wanted to reuse implementation files.

**Comment by yuriks** [comment]
I think including Citra as a monolithic submodule is what I would choose, however I'm not sure if it'll work out: Our headers have a lot of platform-specific cruft inherited from Common in them that probably won't compile in dkA.

Separate submodules in Citra would be terrible, if nothing else because of our PR workflow, and would force splitting of commits.

**Pull request closed without merging**